app/controllers/api/v1x0/stageaction_controller.rb

```diff
@@ -0,0 +1,85 @@
+module Api
+  module V1x0
+    class StageactionController < ActionController::Base
+      include Response
+      include ExceptionHandler
+
+      rescue_from Exceptions::RBACError, Exceptions::ApprovalError, URI::InvalidURIError, ArgumentError do |e|
+        response.body = e.message
+        render :status => :internal_server_error, :action => :result
+      end
+
+      rescue_from ActionController::ParameterMissing do |e|
+        response.body = e.message
+        render :status => :unprocessable_entity, :action => :result
+      end
+
+      before_action :set_stage, :only => [:show, :update]
+
+      def show
+      end
+
+      def update
+        operation = params.require(:commit)
+        case operation
+        when 'Memo', 'Deny'
+          comments = params.require(:message)
+        else
+          comments = params[:message]
+        end
+
+        ManageIQ::API::Common::Request.with_request(@stage.request.context.transform_keys(&:to_sym)) do
+          ActsAsTenant.with_tenant(Tenant.find(@stage.tenant_id)) do
+            ActionCreateService.new(@stage.id).create(
```

Comment:
Right now `ActionCreateService` may raise general `ApprovalError` if the state transition is now allowed. We may need to raise a specific error like `InvalidStateTransitionError`. Here we should catch this error and report a proper error message like `Your action cannot be executed. Somebody else may have approve/denied this request at the mean time.`